Neos.Neos/Classes/Domain/Workspace/WorkspacePublisher.php

```diff
@@ -0,0 +1,283 @@
+<?php
+
+/*
+ * This file is part of the Neos.Neos package.
+ *
+ * (c) Contributors of the Neos Project - www.neos.io
+ *
+ * This package is Open Source Software. For the full copyright and license
+ * information, please view the LICENSE file which was distributed with this
+ * source code.
+ */
+
+declare(strict_types=1);
+
+namespace Neos\Neos\Domain\Workspace;
+
+use Neos\ContentRepository\Core\ContentRepository;
+use Neos\ContentRepository\Core\Feature\WorkspacePublication\Command\DiscardIndividualNodesFromWorkspace;
+use Neos\ContentRepository\Core\Feature\WorkspacePublication\Command\PublishIndividualNodesFromWorkspace;
+use Neos\ContentRepository\Core\Feature\WorkspacePublication\Dto\NodeIdsToPublishOrDiscard;
+use Neos\ContentRepository\Core\Feature\WorkspacePublication\Dto\NodeIdToPublishOrDiscard;
+use Neos\ContentRepository\Core\NodeType\NodeTypeName;
+use Neos\ContentRepository\Core\Projection\ContentGraph\Filter\FindClosestNodeFilter;
+use Neos\ContentRepository\Core\Projection\ContentGraph\VisibilityConstraints;
+use Neos\ContentRepository\Core\Projection\Workspace\Workspace;
+use Neos\ContentRepository\Core\SharedModel\Node\NodeAggregateId;
+use Neos\ContentRepository\Core\SharedModel\Workspace\WorkspaceName;
+use Neos\ContentRepositoryRegistry\ContentRepositoryRegistry;
+use Neos\Flow\Annotations as Flow;
+use Neos\ContentRepository\Core\Feature\WorkspaceRebase\Command\RebaseWorkspace;
+use Neos\ContentRepository\Core\SharedModel\Workspace\ContentStreamId;
+use Neos\Neos\Domain\Service\NodeTypeNameFactory;
+use Neos\Neos\PendingChangesProjection\Change;
+use Neos\Neos\PendingChangesProjection\ChangeFinder;
+
+/**
+ * Neos' workspace publisher service
+ *
+ * @api
+ */
+#[Flow\Scope('singleton')]
+readonly class WorkspacePublisher
+{
+    public function __construct(
+        private ContentRepositoryRegistry $contentRepositoryRegistry
+    ) {
+    }
+
+    /**
+     * @return int The amount of changes that were published
+     */
+    public function publishSite(PublishSite $command): int
+    {
+        $contentRepository = $this->contentRepositoryRegistry->get($command->contentRepositoryId);
+
+        $nodeIdsToPublish = $this->resolveNodeIdsToPublishOrDiscard(
+            $contentRepository,
+            $command->workspaceName,
+            $command->siteId,
+            NodeTypeName::fromString('Neos.Neos:Site')
+        );
+
+        $this->publishNodes(
```

Comment:
Don't we have to check whether the given `$command->siteId` is really a site node?

Reply:
I don't think we really can. If the site's node type was changed to something else than Site, this could never again be discarded. Or am I missing something?

Reply:
the sites nodetype cannot be change because `Neos.Neos:Sites` only allows `Neos.Neos:Site`

Reply:
...until someone changes that.
But I guess this is rather hypothetical and I'm a big fan of constrait checks, so...

Reply:
constraint checks have been added